### PR TITLE
G2: UI surrounding the editor canvas

### PR DESCRIPTION
**Pull request by youknowriad**
bulids on top of #19344 with some additional items that can be harder to land without changing the whole canvas:

 - [x] Dropdowns with darker borders
 - [ ] Updated border-less panels
 - [ ] Darker borders for sidebar, header and footer.

### CODE REVIEW

**Comment by jasmussen** [comment]
I rebased this off G2 (also recently rebased), and started initial work.

But I'm starting slowly with this, because we are getting into some somewhat deep surgery of the editor — a reduction in colors (3 shades of gray instead of 10), a reduction, and other simplification work. For that reason in my first commit I focused only on the popovers in https://github.com/WordPress/gutenberg/pull/19348/commits/ba4fa80c750f313b9ab626c788bffce8f5779432 — yet that still touched a lot of files, so feel free to glance over them to get a flavor for some of the change work we're looking at.

And it's not even done, it's only this far:

Screenshot 2020-01-22 at 12 56 37

This needs padding and spacing alignments, the icons and keyboard shortcuts need to be on the right, and we need a special heuristic for the popover component: most of them should have the light gray border and shadow, but all of the popover menus inside the editing canvas should have the dark border and no shadow. 

Because there are always opinions on this, I don't want to move too fast.

**Comment by jasmussen** [comment]
I took the liberty of renaming this PR. Can we rename the branch also? If it's tricky, no trouble.

**Comment by youknowriad** [comment]
We could rename the branch but I think it would required creating a separate PR. So maybe just keep it as is.

File: packages/components/src/menu-group/style.scss [hunk 1]
```diff
@@ -1,15 +1,12 @@
...
-
 .components-menu-group + .components-menu-group {
 	margin-top: $grid-unit-10;
 	padding-top: $grid-unit-10;
-	border-top: $border-width solid $light-gray-500;
+	border-top: $border-width solid $dark-gray-primary;
 }
 
 .components-menu-group__label {
+	text-transform: uppercase;
```
Comment (ZebulanStanphill) [inline packages/components/src/menu-group/style.scss]: Isn't all-caps considered worse for accessibility?

Comment (jasmussen) [reply packages/components/src/menu-group/style.scss]: I wasn't aware they were. 

I personally think that for a menu group label, which is decidedly seconary (light gray, smaller font size), it's okay to use it as a visual element, it helps the label also act as a separtor. However I'm open to suggestions, any thoughts @pablohoneyhoney?

**Comment by jasmussen** [comment]
Made slight progress on simplifying the toolbar, which includes introducing a basic grid, rearranging the top left buttons, and moving the document outline to the footer (fixes #4287):

Screenshot 2020-01-24 at 12 04 41

**Please do not consider the above representative of anything other than work in progress.**

While the simplificaiton of the toolbar (grid, buttons), work is very far from done, and honestly this reduced border-style makes even clearer that we are in need of icons that fit together better. As the UI is reduced, it increasingly becomes clear that we are using a mix of icons and styles here. To no-ones fault, but a good reason for a fresh singular approach.

File: packages/block-editor/src/components/inserter/style.scss [hunk 2]
```diff
@@ -28,12 +28,24 @@ $block-inserter-search-height: 38px;
...
 	align-items: center;
-	background: none;
 	cursor: pointer;
 	border: none;
 	outline: none;
+	padding: 0;
 	transition: color 0.2s ease;
+
+	// Spot color button.
+	// @todo: This button should ideally be 32px.
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/style.scss]: The inserter is used in different places and not just the top toolbar. I think the "blue button" shouldn't be solved using CSS but using the `isPrimary` prop in JS and only for the top toolbar inserter.

**Comment by ZebulanStanphill** [comment]
One thing that has always confused me is that the <kbd><samp>Save Draft</samp></kbd> and <kbd><samp>Preview</samp></kbd> buttons use different styles. I would think that they should look the same.

Also, should the Document Outline button use a visible label? I would think that there is enough space for it. I'm also not convinced it belongs in the footer. Perhaps it should be an extension similar to what Yoast and other plugins have, with the ability to pin it to the right-side of the top toolbar.

Finally, I wonder if the Block Navigation menu belongs in the footer instead of the Document Outline. It has a similar function to the breadcrumb links. Or maybe they both belong there.

**Comment by jasmussen** [comment]
> One thing that has always confused me is that the Save Draft and Preview buttons use different styles. I would think that they should look the same.

I still hope that we can entirely deprecate Save Draft as a vestige of the past. However it's increasingly clear that this is still a ways out in the future. In the mean time, yes, it does make sense to unify on a single button style. Though it is worth noting that there might land a different interface for the Preview button, when the editor [gets responsive features and registers an editor style](https://github.com/WordPress/gutenberg/pull/19082#issuecomment-567508590).

> Also, should the Document Outline button use a visible label? I would think that there is enough space for it. I'm also not convinced it belongs in the footer. Perhaps it should be an extension similar to what Yoast and other plugins have, with the ability to pin it to the right-side of the top toolbar.

IMO, it should be a word count visible label, yes. Like so:

Screenshot 2020-01-27 at 10 10 38

Then you click it, and the document outline interface shows up. The icon isn't particularly helpful there, but a label showing real time information combined with a tooltip feels like a good balance to me. 

> Finally, I wonder if the Block Navigation menu belongs in the footer instead of the Document Outline. It has a similar function to the breadcrumb links. Or maybe they both belong there.

Possibly. While I think the Selection mode and breadcrumb interfaces both have made strides in selecting child blocks, I don't think we'll know how efficient those two are until we actually ship the G2 efforts: specifically the changes to what has focus and what is selected, in my experience of working on the branches so far, make it much clearer what goes on.

That's a roundabout way of saying that until we feel confident enough that those two interfaces work for selecting parent blocks, perhaps we should keep the block navigation menu where it is? And if they prove themselves, your points do make sense that the button could serve as an extension/expansion of the breadcrumbs.

**Comment by ZebulanStanphill** [comment]
> > Also, should the Document Outline button use a visible label? I would think that there is enough space for it. I'm also not convinced it belongs in the footer. Perhaps it should be an extension similar to what Yoast and other plugins have, with the ability to pin it to the right-side of the top toolbar.
> 
> IMO, it should be a word count visible label, yes. Like so:
> 
> Screenshot 2020-01-27 at 10 10 38
> 
> Then you click it, and the document outline interface shows up. The icon isn't particularly helpful there, but a label showing real time information combined with a tooltip feels like a good balance to me.

I agree that the icon isn't that helpful in the footer, but I don't think the word count makes for a good button to open up the Document Outline, since it is not a description of what the button does. We would have to use hidden labels on screen readers, and for everyone else it wouldn't be clear that word count is a button in the first place.

**Comment by jasmussen** [comment]
> I don't think the word count makes for a good button to open up the Document Outline, since it is not a description of what the button does.

I actually quite disagree with this — the button shows a word count and shows word-count related information when clicked. 

I do agree it could use additional sanity checking, so I'll add a label and invite that.

**Comment by jasmussen** [comment]
In this branch I refactored how hover and focus styles are specified to greatly reduce specificity. So far it's working, and as a result a ton of extra CSS can be removed and simplified. But it needs testing in all edgecases, so look for hover and focus to work in every situation.

Unified the add block buttons. It really deserves a new icon soon. 

Screenshot 2020-02-06 at 12 25 24

**Comment by ZebulanStanphill** [comment]
The empty Paragraph and placeholder inserters haven't changed. Is that intentional?
image

Accordions still use the default focus style. Is that intentional?
image

Accordions also don't have blue text when you hover them. In fact, they don't have any hover styles at all.
image

The sidebar tabs don't use blue text on hover either.
image

There seems to be a bit too much space at the bottom of the <kbd><samp>More tools & options</samp></kbd> menu.
image

These buttons don't line up horizontally, even though you'd think they would. There's not much padding in the sidebar compared to the top bar.
image

<kbd><samp>Remove from Reusable blocks</samp></kbd> still has an icon.
image

The focus border is thicker for the select control than for other UI elements. Also, the text in these controls seems kind of small. Is that intentional?
image

Hovering the inserter button makes it lighter. Hovering the Publish button makes it darker.
image
image

**Comment by jasmussen** [comment]
Excellent feedback. I haven't gotten to all of that yet, and it is my intention to do so, and your screenshots are still helpful!

**Comment by jasmussen** [comment]
Got a lot of things done today. 

One thing I'd like to surface is https://github.com/WordPress/gutenberg/pull/19348/commits/410befdf226298e8cb67e513fa821d77cc072289, which makes the separator easier to select. I could swear there was an issue for this, but perhaps I'm misremembering?

In any case, it now leverages the block margins to make the thing selectable. Perhaps it's the lack of block borders and block paddings that makes this a non issue, compared to the past. 

Frontend:

front

Editor:

editor

**Comment by jasmussen** [comment]
Polished top toolbar and simplified the CSS a bit. The mover control is now the same.

top toolbar

Works horizontally as well of course:

Screenshot 2020-02-10 at 11 42 33

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/pull/19348/commits/8006ab0158c77288a6a2ee6766fca7734cde1a7d I polished the placeholder component.

I adjusted a font size, some paddings, and removed the icon. This G2 effort overall removes quite a few icons in many places, including dropdown menus. This is especially useful when placeholders are shown in smaller contexts, but brings a bit consistency across those dimensions:

Screenshot 2020-02-10 at 12 50 30

Note, those hanging plus icons are not unique to this branch, they are showing because of a rule (which we should revisit) of showing the appender after any non-text block. Even when the block is deselected. It's not a good rule.

**Comment by jasmussen** [comment]
I polished menus and the block switcher ever so slightly, and I discovered this:

popover direction

Popovers open in lots of different direction. Why? I understand there's some math going on to make sure it opens in a direction that has sufficient space for it to be visible, i.e. don't open to the right if you're close to the right edge of the screen. But in the GIF example above, there's room in the same direction for all the menus, so why don't they all open from top left and in the direction of down and right?

**Comment by ZebulanStanphill** [comment]
@jasmussen

The "Add block" button is the same color as the focus border, and there's no separation between the two, so you can't immediately tell if it is focused or not.

I agree that the label and description could be merged to say something simple like "Insert image" or "Insert media" (in the case of the Media & Text block). Also, I think it is a bit inconsistent to have one button labeled "Media Library" and another labeled "Insert from URL"; shouldn't they both start (or not start) with "Insert from"?

**Comment by jasmussen** [comment]
Great thoughts, good catches. Lots to do, I'll get there.

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/pull/19348/commits/0a74609fc281dbb4d0f327885affd33bd9b428e8, I started work on the Button component refactor:

buttons

This will need more work, but it's a great way to test the ideas outlined in #18667 in actual practice. I initially thought it was going to be an issue to remove the border around the secondary button, but in practice so far, it feels like the value that the fewer borders brings to clarity and hierarchy has substantial value, and the context makes it obvious what's button, just like how buttons in a dropdown menu still obviously remain dropdown menu items. 

We'll need more work here, on margins, paddings, font weight and heuristics. But it's a good time to test so far.

**Comment by ZebulanStanphill** [comment]
Things are really feeling a lot more lightweight compared to `master`. :+1:

Some more observations as of the latest changes:

The inline text color indicator doesn't quite work with the current toolbar button focus style.
image

The revisions link seems to be lacking some margin/padding.
image

The block style/transform popover has a bunch of padding now. Is that intentional?
image

The Save/switch to draft button changes text color slightly on hover, but the Preview button does not.
image

The subtle background when you press the Preview button is not present for the Save/switch to draft button. Icon-only buttons also lack this style.
image

Additionally, it seems a bit odd that this style is being used for a pressed state for secondary text buttons, but elsewhere it is used as a hover style. That seems inconsistent to me.
image
image

Meanwhile, the Block navigation menu uses it to show the currently-selected block, which aligns more with the current usage for the secondary text buttons.
image

I'm not sure how to solve this, but one big issue here is that blue is used as the foreground/text color for icon buttons, but text-only buttons use it as the default text color, so you can't use the same hover style for text buttons and icon buttons.

The answer might be to make text buttons use a black text color (just like dropdown menus), but I worry that might make it too hard to distinguish what is a button and what is just text. What do you think? It's worth noting that the "Replace" button in the block toolbar is already black.

The opposite direction, making all icon buttons blue by default, seems like it would mess with the contrast too much, so I think that's out of the question.

**Comment by jasmussen** [comment]
Incredible feedback as usual. I'm out for the day but rest assured I'll work through it!

**Comment by jasmussen** [comment]
I pushed some fixes, including to the text color, polish to the transform dialog, and better code for the buttons. 

Your point on the hover vs. selected states is well made. I suspect the buttons need a little more attention, and certainly feedback. Visuals speak better so here are a couple of GIFs:

icons

header

nav

toggled

switcher

Of those, to me the heuristics feel sensible (if able to receive more polish) with regards to hover, active and focus styles. The total inversion for toggled and selected states feels very powerful in every situation except perhaps the style picker, where it feels too much. I wonder if that's a matter of unifying the border radius or overall looking at a different treatment. But for now it felt important to try it out in principle.

**Comment by jasmussen** [comment]
Polished the gallery:

gallery

This block deserves a refactor, separately. Notably we had this idea:

Screenshot 2020-02-12 at 10 01 27

Increasingly this seems like a less good idea, because it separates the gallery from the caption below. It seems like with the "Replace" button landing on Images, a similar one could be added to the gallery.

**Comment by jasmussen** [comment]
Alright — I finished that last item on my list. There's a lot still to do, but some of that I've so far cautiously put on a list called "followup". That list includes some non-trivial things such as the sidebar and other aspects, which obviously need addressing. However I'm trying to find a balance here between keeping the already-substantial change this G2 effort is at a place where _enough is addressed that a proper feel for the change can be had_, while at the same time keeping the pull requests from growing any bigger. 

My hope is that this balance lets us act on feedback this branch can receive in the coming weeks (and for as long as necessary), and hopefully result in proving the overall concept deserving of being merged into the master branch not too long after the 5.4 cycle has ended. Provided that is possible, it should give us ample time to address the items so far unaddressed.

**Comment by ZebulanStanphill** [comment]
I think this branch is already at a really good state. I think a lot of the remaining issues can be handled with later. There are a few things I would like to see addressed before a merge, though:

When clicking a block, a light gray overlay appears briefly before fading away. Is that intentional?

Are we sure we want to get rid of the accordions in the block style/transform switcher?

The block toolbar ellipsis dropdown is positioned a bit too far down when it opens downwards. It's fine when opening upwards, though. Also, the Remove from Reusable blocks button still needs to have its icon removed.
image

**Comment by jasmussen** [comment]
I had meant to polish a bit further today, but landed massive cold so I'll have to stay bedridden for the day.

> When clicking a block, a light gray overlay appears briefly before fading away. Is that intentional?

This is an experiment, yes, to briefly indicate the footprint of the block, try and find a balance between _this is an editor that doesn't need borders around everything_, and addressing the specific issue of the toolbar sitting far on the left of a right-aligned short paragraph. It's one of those things that is best tested in the branch, I believe.

> Are we sure we want to get rid of the accordions in the block style/transform switcher?

Eventually, probably. For this v1 branch? Possibly not? In removing them I had hoped to take the first step towards this idea:

Drop downs

That idea is one of reducing the overall footprint of the switcher and simplifying, while still hopefully making styles more useful. It will no doubt require more dev work and iteration separately, but there's going to be a phase where we're in-between. Open to suggestions here. 

> The block toolbar ellipsis dropdown is positioned a bit too far down when it opens downwards. It's fine when opening upwards, though.

I have some frustrations with how this is currently working, as was noted in https://github.com/WordPress/gutenberg/pull/19348#issuecomment-584100911. 

Interestingly, this is one of those discrepancies that became visible in this branch, but I never thought about it in master — I feel that's a good sign, we've spent a lot of time fixing bugs and growing the featureset and it organically grew into what's shipping. But when giving an intentional focus to the visuals, suddenly a lot of inconsistencies and cruft appear.

I'm not sure how much dev work is required to get the popovers to open consistently — as shown in the switcher screen above, I'd like to try the interface where popovers from the block toolbar are shown in the same visual treatment and with the same spacing between elements. And when there's space, it feels like the dropdowns should all open in the same direction. I'll see if I can make the spacing consistent, at least.

**Comment by ZebulanStanphill** [comment]
> I had meant to polish a bit further today, but landed massive cold so I'll have to stay bedridden for the day.

Hope you get well soon! I just recently got over being sick myself.

> > When clicking a block, a light gray overlay appears briefly before fading away. Is that intentional?
> 
> This is an experiment, yes, to briefly indicate the footprint of the block, try and find a balance between _this is an editor that doesn't need borders around everything_, and addressing the specific issue of the toolbar sitting far on the left of a right-aligned short paragraph. It's one of those things that is best tested in the branch, I believe.

Well, it seems to work alright. I'm not really sure how effective it is, but it doesn't seem to have any negative effects as far as I can tell.

> > Are we sure we want to get rid of the accordions in the block style/transform switcher?
> 
> Eventually, probably. For this v1 branch? Possibly not? In removing them I had hoped to take the first step towards this idea:
> 
> Drop downs

I wonder how those mockups would work on mobile? Sideways-opening menus tend to have issues on mobile devices, while accordions work fine on all platforms.

**Comment by jasmussen** [comment]
Rebased the try/g2 branch, reset this branch to that, cherry picked, verified it ran, then squashed this to make future cherrypicks easier. Hopefully everything works as it should still, it seems to for me :D 

I'll now proceed to adress all the great feedback that's been piling up.

**Comment by jasmussen** [comment]
> I wonder how those mockups would work on mobile? Sideways-opening menus tend to have issues on mobile devices, while accordions work fine on all platforms.

I think this is good feedback, and something you should consider sharing on the main ticket itself: #18667. I can think of a number of ways to address that, including simply behaving as a sequence of dialogs (like when you click a dropdown on your phone and it opens a dialog). But it's both something that deserves more thought, and will not be addressed immediately in this PR, so totally worth surfacing.

**Comment by jasmussen** [comment]
This PR fixes #20254.

**Comment by youknowriad** [comment]
Capture d’écran 2020-02-19 à 11 36 21 AM

I tried updating the button height to 36px to keep it consistent with the other buttons in the top bar. It looks way better for me. Also, I learned that everytime we create a specific inconsistency, we kind of go back later, so my preference would be to stay consistent here.

It doesn't work that well for the sibling inserter but I believe the sibling inserter doesn't work at all as a "primary" button (blue).

**Comment by jasmussen** [comment]
Sibling inserter in 36x36 for reference:

Screenshot 2020-02-19 at 11 40 37

**Comment by youknowriad** [comment]
I noticed that the panel header (block | document) in the sidebar is now white, I think ultimately that's a good change but right now, because the sub panels (inspector panels and document settings) share the same design (white with light gray border), I feel it removes some hierarchy. So I wonder if we should do it now or only when we revisit the design of these panels.

**Comment by jasmussen** [comment]
White sidebar tabs, for reference:

Screenshot 2020-02-19 at 15 14 36

Happy to address based on any consensus, and I'm eager to see if we can merge this into G2 soon and be that much closer 🤞

**Comment by jasmussen** [comment]
I rebased. Here's a little taste test to whet the appetites for a merge into g2:

g3

I noticed that when the Top Toolbar option is enabled, the movers still do the smart unhiding. We should probably just toggle that off up there. @ItsJonQ can you help with that? Can commit straight to the try/g2 branch now :)

**Comment by ItsJonQ** [comment]
@jasmussen Taking a look when I can :D

**Comment by ItsJonQ** [comment]
@jasmussen Haiii! The Movers seem to be behaving on my end :o. I've been resizing desktop <-> mobile, as well as loading the page at various viewport sizes.

In doing so, I discovered this:

Screen Shot 2020-02-19 at 9 48 00 AM

Not sure if this already exists. It's when the viewport is **exactly** 782px

**Comment by jasmussen** [comment]
Documenting to fix either in this branch or in g2 or in a followup: the block toolbar when seen in one of the responsive views (from the new dropdown) becomes a broken mix of what it is in the top toolbar and a by-the-block version.

**Comment by jasmussen** [comment]
Q can you clarify what the issue is in that screenshot, it's hard to tell?

Also, what I mean about the top toolbar is when it's permanently on:

Screenshot 2020-02-19 at 15 53 25

In that case the movers keep their appear behavior:

Screenshot 2020-02-19 at 15 53 32

It seems to work fine on the mobile breakpoints where the toolbar is forced up there.

**Comment by ZebulanStanphill** [comment]
@youknowriad
> It doesn't work that well for the sibling inserter but I believe the sibling inserter doesn't work at all as a "primary" button (blue) because it's not a primary action anyway.

I agree; the sibling inserter shouldn't be considered a primary button.

**Comment by jasmussen** [comment]
Good thoughts. 

I think many of these thoughts are based on the half-step, even if it's still a big step, that this branch is. The sidebar needs a LOT of love, and a white header takes a step towards that. But I'm fine to explore that in a followup, so it's now gray again, even if that looks out of place with the reduced color set — (most of the rest of the sidebar also looks out of place :D ):

Screenshot 2020-02-20 at 09 48 14

I also made the block library button in the toolbar into an isPrimary button, which allowed me to remove a bunch of CSS. I kept it 32x32 as it's part of aligning those buttons to a 12px grid. I'm happy to also revisit this in a followup and either change that to 36, or explore the 32 in other places.

By making the inserter a primary, ironically, it also made the sibling inserter primary, which meant I had to undo some of those visuals to make it appear not primary:

Screenshot 2020-02-20 at 09 48 20

I don't have strong feelings for the current state of the sibling inserter, I profoundly thing that inserters inside the canvas deserve a whole iteration on its own — can we remove a bunch of them, for example? On the block library ticket that [discusses opening the block library as a modal dialog that defaults to the "Patterns" tab](https://github.com/WordPress/gutenberg/issues/17335#issuecomment-562049752), we made it blue, like so:

Basic Editor

But given it needs more love, and it's holding back this PR, absolutely something to follow up on.

**Comment by youknowriad** [comment]
Thanks for the updates @jasmussen that's great

> By making the inserter a primary, ironically, it also made the sibling inserter primary, which meant I had to undo some of those visuals to make it appear not primary:

I believe we need to have a prop in the inserter to choose whether we want a primary button or not to avoid having to revert the changes in CSS. For example in appenders... I'll see if I can make this change.

**Comment by youknowriad** [comment]
Now, that this branch is very close to G2 and without touching the panels... I think we can merge it with G2 and continue on a single branch. Nice work @jasmussen

**Comment by jasmussen** [comment]
ƪ(˘⌣˘)┐ ƪ(˘⌣˘)ʃ ┌(˘⌣˘)ʃ

**Pull request merged**